File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/review_claims/ReviewClaimsContent.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.review_claims

import app.pantopus.android.data.api.models.admin.AdminClaimBucket
import app.pantopus.android.data.api.models.admin.AdminClaimDto
import app.pantopus.android.data.api.models.admin.AdminClaimHomeDto
import app.pantopus.android.ui.components.StatusChipVariant
import app.pantopus.android.ui.screens.shared.list_of_rows.GradientPair
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Helper types + projections shared by `ReviewClaimsViewModel` and
 * `ReviewClaimDetailViewModel`. Mirrors iOS `ReviewClaimsContent.swift`
 * — same chip / method / evidence-label vocabulary so the same words
 * surface across iOS, Android, and web.
 */

/** Reviewer-facing label for a claim's `method` field. */
object AdminClaimMethodLabel {
    fun display(method: String?): String =
        when (method) {
            "doc_upload" -> "Document Upload"
            "escrow_agent" -> "Escrow/Title Agent"
            "property_data_match" -> "ID Verification"
            "invite" -> "Invited"
            "vouch" -> "Vouched"
            "landlord_portal" -> "Landlord Portal"
            null -> "Unknown method"
            else -> humanize(method)
        }

    private fun humanize(snake: String): String =
        snake
            .replace('_', ' ')
            .split(' ')
            .joinToString(" ") { word ->
                word.replaceFirstChar { it.uppercase() }
            }
}

/** Reviewer-facing label for a piece of `HomeVerificationEvidence`. */
object AdminClaimEvidenceLabel {
    fun display(type: String): String =
        when (type) {
            "deed" -> "Deed"
            "closing_disclosure" -> "Closing Disclosure"
            "tax_bill" -> "Tax Bill"
            "utility_bill" -> "Utility Bill"
            "lease" -> "Lease Agreement"
            "escrow_attestation" -> "Escrow Attestation"
            "title_match" -> "Title Match"
            "idv" -> "ID Verification"
            else ->
                type
                    .replace('_', ' ')
                    .split(' ')
                    .joinToString(" ") { word -> word.replaceFirstChar { it.uppercase() } }
        }
}

/**
 * Six-stop avatar gradient palette keyed by a stable hash of the
 * claimant id. Drawn from [PantopusColors] tokens — no hex literals
 * here.
 */
object AdminClaimAvatarGradient {
    private val palette: List<GradientPair> =
        listOf(
            GradientPair(PantopusColors.primary500, PantopusColors.primary700),
            GradientPair(PantopusColors.error, PantopusColors.business),
            GradientPair(PantopusColors.warning, PantopusColors.handyman),
            GradientPair(PantopusColors.success, PantopusColors.home),
            GradientPair(PantopusColors.business, PantopusColors.primary700),
            GradientPair(PantopusColors.primary600, PantopusColors.primary500),
        )

    fun gradient(seed: String): GradientPair {
        var hash = 0
        for (ch in seed) hash += ch.code
        val idx = (hash % palette.size + palette.size) % palette.size
        return palette[idx]
    }
}

/** Triage chip descriptor — text + icon + variant. */
data class AdminClaimChipDescriptor(
    val text: String,
    val icon: PantopusIcon,
    val variant: StatusChipVariant,
)

/** Build the triage / terminal chip for a row given its bucket + state. */
object AdminClaimChip {
    private const val SECONDS_PER_DAY = 86_400L
    private const val AGING_THRESHOLD_DAYS = 7L

    fun descriptor(
        claim: AdminClaimDto,
        bucket: AdminClaimBucket,
        referenceInstant: Instant = Instant.now(),
    ): AdminClaimChipDescriptor =
        when (bucket) {
            AdminClaimBucket.Approved ->
                AdminClaimChipDescriptor(
                    text = "Approved",
                    icon = PantopusIcon.CheckCircle,
                    variant = StatusChipVariant.Success,
                )
            AdminClaimBucket.Rejected ->
                AdminClaimChipDescriptor(
                    text = "Rejected",
                    icon = PantopusIcon.CircleSlash,
                    variant = StatusChipVariant.ErrorVariant,
                )
            AdminClaimBucket.Pending -> pendingDescriptor(claim, referenceInstant)
        }

    private fun pendingDescriptor(
        claim: AdminClaimDto,
        referenceInstant: Instant,
    ): AdminClaimChipDescriptor {
        if (claim.state == "disputed") {
            return AdminClaimChipDescriptor(
                text = "Conflict",
                icon = PantopusIcon.AlertTriangle,
                variant = StatusChipVariant.ErrorVariant,
            )
        }
        if (claim.state == "needs_more_info") {
            return AdminClaimChipDescriptor(
                text = "Awaiting docs",
                icon = PantopusIcon.Hourglass,
                variant = StatusChipVariant.Neutral,
            )
        }
        val ageDays = ageDays(claim.createdAt, referenceInstant)
        return if (ageDays != null && ageDays >= AGING_THRESHOLD_DAYS) {
            AdminClaimChipDescriptor(
                text = "Aging · ${ageDays}d",
                icon = PantopusIcon.Clock,
                variant = StatusChipVariant.Warning,
            )
        } else {
            AdminClaimChipDescriptor(
                text = "New",
                icon = PantopusIcon.Sparkles,
                variant = StatusChipVariant.Info,
            )
        }
    }

    private fun ageDays(
        iso: String,
        referenceInstant: Instant,
    ): Long? =
        parseInstant(iso)?.let { created ->
            val seconds = (referenceInstant.epochSecond - created.epochSecond).coerceAtLeast(0)
            seconds / SECONDS_PER_DAY
        }
}

/** Human-readable "filed Xago" — short units to match the web row. */
object AdminClaimTimeFormat {
    private const val SECONDS_PER_MINUTE = 60L
    private const val SECONDS_PER_HOUR = 3_600L
    private const val SECONDS_PER_DAY = 86_400L

    fun submittedAgo(
        iso: String,
        referenceInstant: Instant = Instant.now(),
    ): String {
        val instant = parseInstant(iso) ?: return ""
        val seconds = (referenceInstant.epochSecond - instant.epochSecond).coerceAtLeast(0)
        return when {
            seconds < SECONDS_PER_MINUTE -> "filed just now"
            seconds < SECONDS_PER_HOUR -> "filed ${seconds / SECONDS_PER_MINUTE}m ago"
            seconds < SECONDS_PER_DAY -> "filed ${seconds / SECONDS_PER_HOUR}h ago"
            else -> "filed ${seconds / SECONDS_PER_DAY}d ago"
        }
    }

    /** Format the banner's oldest-in-queue subtitle. */
    fun oldestAge(seconds: Int?): String {
        if (seconds == null) return "no claims"
        return when {
            seconds < SECONDS_PER_MINUTE -> "${seconds}s"
            seconds < SECONDS_PER_HOUR -> "${seconds / SECONDS_PER_MINUTE} min"
            seconds < SECONDS_PER_DAY -> "${seconds / SECONDS_PER_HOUR}h"
            else -> "${seconds / SECONDS_PER_DAY}d"
        }
    }

    /** Format an ISO date as "Mar 4, 2026". */
    fun longDate(iso: String): String {
        val instant = parseInstant(iso) ?: return iso
        return runCatching {
            LONG_DATE_FMT.format(instant.atZone(ZoneId.systemDefault()))
        }.getOrDefault(iso)
    }

    private val LONG_DATE_FMT: DateTimeFormatter = DateTimeFormatter.ofPattern("MMM d, yyyy")
}

/** Address line projection for the row subtitle + detail header. */
object AdminClaimAddressFormat {
    fun full(home: AdminClaimHomeDto?): String {
        if (home == null) return "Unknown address"
        val head = home.name?.takeIf { it.isNotEmpty() } ?: home.address
        val parts =
            listOfNotNull(
                head?.takeIf { it.isNotEmpty() },
                home.city?.takeIf { it.isNotEmpty() },
                home.state?.takeIf { it.isNotEmpty() },
            )
        return if (parts.isEmpty()) "Unknown address" else parts.joinToString(", ")
    }
}

/** Parse common ISO-8601 variants (with / without fractional seconds, with / without offset). */
internal fun parseInstant(iso: String): Instant? {
    if (iso.isBlank()) return null
    runCatching { return Instant.parse(iso) }
    // Naive timestamps (no trailing offset) — assume UTC.
    runCatching { return Instant.parse(iso + "Z") }
    return null
}

File: frontend/apps/ios/Pantopus/Features/ReviewClaims/ReviewClaimsContent.swift
//
//  ReviewClaimsContent.swift
//  Pantopus
//
//  Helper types + projections shared by `ReviewClaimsViewModel` and
//  `ReviewClaimDetailViewModel`. Mirrors the web client's per-bucket
//  chip / method / evidence-label lookup tables so the same vocabulary
//  reads identically across iOS, Android, and web.
//

import Foundation
import SwiftUI

/// Reviewer-facing label for a claim's `method` field. Mirrors the
/// `METHOD_LABELS` map in the web review-claims page.
public enum AdminClaimMethodLabel {
    public static func display(for method: String?) -> String {
        switch method {
        case "doc_upload": "Document Upload"
        case "escrow_agent": "Escrow/Title Agent"
        case "property_data_match": "ID Verification"
        case "invite": "Invited"
        case "vouch": "Vouched"
        case "landlord_portal": "Landlord Portal"
        case let other?: humanize(other)
        case nil: "Unknown method"
        }
    }

    private static func humanize(_ snake: String) -> String {
        snake
            .replacingOccurrences(of: "_", with: " ")
            .capitalized
    }
}

/// Reviewer-facing label for a piece of `HomeVerificationEvidence`.
/// Mirrors the `EVIDENCE_LABELS` map in the web review-claims page.
public enum AdminClaimEvidenceLabel {
    public static func display(for type: String) -> String {
        switch type {
        case "deed": "Deed"
        case "closing_disclosure": "Closing Disclosure"
        case "tax_bill": "Tax Bill"
        case "utility_bill": "Utility Bill"
        case "lease": "Lease Agreement"
        case "escrow_attestation": "Escrow Attestation"
        case "title_match": "Title Match"
        case "idv": "ID Verification"
        default:
            type
                .replacingOccurrences(of: "_", with: " ")
                .capitalized
        }
    }
}

/// Six-stop avatar gradient palette keyed by a stable hash of the
/// claimant id. Drawn from `Theme.Color` tokens — no hex literals here
/// (the underlying colour values live in `Theme.Color`).
public enum AdminClaimAvatarGradient {
    public static func gradient(for seed: String) -> GradientPair {
        let palette: [GradientPair] = [
            GradientPair(start: Theme.Color.primary500, end: Theme.Color.primary700),
            GradientPair(start: Theme.Color.error, end: Theme.Color.business),
            GradientPair(start: Theme.Color.warning, end: Theme.Color.handyman),
            GradientPair(start: Theme.Color.success, end: Theme.Color.home),
            GradientPair(start: Theme.Color.business, end: Theme.Color.primary700),
            GradientPair(start: Theme.Color.primary600, end: Theme.Color.primary500)
        ]
        let hash = seed.unicodeScalars.reduce(0) { $0 &+ Int($1.value) }
        return palette[abs(hash) % palette.count]
    }
}

/// Triage chip rendered on a row inside the Pending tab. The Approved /
/// Rejected buckets get a terminal-state chip instead.
public struct AdminClaimChipDescriptor {
    public let text: String
    public let icon: PantopusIcon
    public let variant: StatusChipVariant
}

public enum AdminClaimChip {
    /// Build the trailing status chip for a row given its bucket + state.
    public static func descriptor(
        for claim: AdminClaimDTO,
        bucket: AdminClaimBucket,
        referenceDate: Date = Date()
    ) -> AdminClaimChipDescriptor {
        switch bucket {
        case .approved:
            return AdminClaimChipDescriptor(text: "Approved", icon: .checkCircle, variant: .success)
        case .rejected:
            return AdminClaimChipDescriptor(text: "Rejected", icon: .circleSlash, variant: .error)
        case .pending:
            break
        }

        if claim.state == "disputed" {
            return AdminClaimChipDescriptor(text: "Conflict", icon: .alertTriangle, variant: .error)
        }
        if claim.state == "needs_more_info" {
            return AdminClaimChipDescriptor(text: "Awaiting docs", icon: .hourglass, variant: .neutral)
        }

        let parser = ISO8601DateFormatter()
        parser.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = parser.date(from: claim.createdAt)
            ?? ISO8601DateFormatter().date(from: claim.createdAt)
        let ageSeconds = date.map { referenceDate.timeIntervalSince($0) } ?? 0
        let ageDays = Int(ageSeconds / 86400)
        if ageDays >= 7 {
            return AdminClaimChipDescriptor(
                text: "Aging · \(ageDays)d",
                icon: .clock,
                variant: .warning
            )
        }
        return AdminClaimChipDescriptor(text: "New", icon: .sparkles, variant: .info)
    }
}

/// Human-readable "Submitted x ago" — short units to match the web row.
public enum AdminClaimTimeFormat {
    public static func submittedAgo(
        _ iso: String,
        referenceDate: Date = Date()
    ) -> String {
        let parser = ISO8601DateFormatter()
        parser.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = parser.date(from: iso) ?? ISO8601DateFormatter().date(from: iso) else {
            return ""
        }
        let seconds = max(0, Int(referenceDate.timeIntervalSince(date)))
        switch seconds {
        case ..<60: return "filed just now"
        case ..<3600: return "filed \(seconds / 60)m ago"
        case ..<86400: return "filed \(seconds / 3600)h ago"
        default: return "filed \(seconds / 86400)d ago"
        }
    }

    /// Format the banner's oldest-in-queue subtitle (e.g. "Oldest in
    /// queue: 5h").
    public static func oldestAge(_ seconds: Int?) -> String {
        guard let s = seconds else { return "no claims" }
        if s < 60 { return "\(s)s" }
        if s < 3600 { return "\(s / 60) min" }
        if s < 86400 { return "\(s / 3600)h" }
        return "\(s / 86400)d"
    }

    /// Format a posix date as "Mar 4, 2026" — used in the detail summary.
    public static func longDate(_ iso: String) -> String {
        let parser = ISO8601DateFormatter()
        parser.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = parser.date(from: iso) ?? ISO8601DateFormatter().date(from: iso) else {
            return iso
        }
        return Self.longFormatter.string(from: date)
    }

    private static let longFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM d, yyyy"
        return f
    }()
}

/// Static formatting helpers for the row subtitle (address) and meta tail.
public enum AdminClaimAddressFormat {
    public static func full(_ home: AdminClaimHomeDTO?) -> String {
        guard let home else { return "Unknown address" }
        let head = home.name?.isEmpty == false ? home.name : home.address
        let parts = [head, home.city, home.state].compactMap { $0?.isEmpty == false ? $0 : nil }
        return parts.isEmpty ? "Unknown address" : parts.joined(separator: ", ")
    }
}
